plugin/core/tree_view.py

```diff
@@ -0,0 +1,269 @@
+            disclosure_button_html = '<a class="disclosure-button" href="{}">▶</a>'.format(
+                sublime.command_url('lsp_expand_tree_item', {'name': sheet_name, 'id': self.id}))
+        elif self.collapsible_state == TreeItemCollapsibleState.EXPANDED:
+            disclosure_button_html = '<a class="disclosure-button" href="{}">▼</a>'.format(
+                sublime.command_url('lsp_collapse_tree_item', {'name': sheet_name, 'id': self.id}))
+        else:
+            disclosure_button_html = '<span class="disclosure-button">&nbsp;</span>'
+        icon_html = '<span class="{}" title="{}">{}</span>'.format(
+            self._kind_class_name(self.kind[0]), self.kind[2], self.kind[1] if self.kind[1] else '&nbsp;')
+        if self.command_url and self.tooltip:
+            label_html = '<a class="label" href="{}" title="{}">{}</a>'.format(
+                self.command_url, html.escape(self.tooltip), html.escape(self.label))
+        elif self.command_url:
+            label_html = '<a class="label" href="{}">{}</a>'.format(self.command_url, html.escape(self.label))
+        elif self.tooltip:
+            label_html = '<span class="label" title="{}">{}</span>'.format(
+                html.escape(self.tooltip), html.escape(self.label))
+        else:
+            label_html = '<span class="label">{}</span>'.format(html.escape(self.label))
+        description_html = '<span class="description">{}</span>'.format(html.escape(self.description)) if \
+            self.description else ''
+        return '<div class="tree-view-row">{}</div>'.format(
+            indent_html + disclosure_button_html + icon_html + label_html + description_html)
+
+    @staticmethod
+    def _kind_class_name(kind_id: int) -> str:
+        if kind_id == sublime.KIND_ID_KEYWORD:
+            return "kind kind_keyword"
+        if kind_id == sublime.KIND_ID_TYPE:
+            return "kind kind_type"
+        if kind_id == sublime.KIND_ID_FUNCTION:
+            return "kind kind_function"
+        if kind_id == sublime.KIND_ID_NAMESPACE:
+            return "kind kind_namespace"
+        if kind_id == sublime.KIND_ID_NAVIGATION:
+            return "kind kind_navigation"
+        if kind_id == sublime.KIND_ID_MARKUP:
+            return "kind kind_markup"
+        if kind_id == sublime.KIND_ID_VARIABLE:
+            return "kind kind_variable"
+        if kind_id == sublime.KIND_ID_SNIPPET:
+            return "kind kind_snippet"
+        return "kind kind_ambiguous"
+
+
+class Node:
+
+    __slots__ = ('element', 'tree_item', 'indent_level', 'children', 'is_resolved')
+
+    def __init__(self, element: T, tree_item: TreeItem, indent_level: int = 0) -> None:  # pyright: ignore
+        self.element = element
+        self.tree_item = tree_item
+        self.indent_level = indent_level
+        self.children = []  # type: List[str]  # IDs of child nodes
+        self.is_resolved = False
+
+
+class TreeDataProvider(metaclass=ABCMeta):
```

Comment:
I believe it would be a bit more correct from types standpoint (I've noticed some mypy errors in subclass) to use `class TreeDataProvider(Generic[T]):` but then we couldn't take advantage of `@abstractmethod`. Or at least I don't see a way to use both.